soc/st/stm32/stm32wbax/power.c

```diff
@@ -18,69 +23,71 @@
 #include <stm32wbaxx_ll_system.h>
 #include <clock_control/clock_stm32_ll_common.h>
 
-#ifdef CONFIG_BT_STM32WBA
-#include "scm.h"
-#endif
-
 #include <zephyr/logging/log.h>
 
 LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
 
-void stm32_power_init(void);
 
-static void set_mode_stop(uint8_t substate_id)
-{
+#define HSE_ON (READ_BIT(RCC->CR, RCC_CR_HSEON) == RCC_CR_HSEON)
 
-	LL_PWR_ClearFlag_STOP();
-	LL_RCC_ClearResetFlags();
+static uint32_t ram_waitstates_backup;
+static uint32_t flash_latency_backup;
 
-	/* Erratum 2.2.15:
-	 * Disabling ICACHE is required before entering stop mode
-	 */
-	sys_cache_instr_disable();
-
-#ifdef CONFIG_BT_STM32WBA
-	scm_setwaitstates(LP);
+#if defined(CONFIG_PM_S2RAM)
+static struct fpu_ctx_full fpu_state;
+static struct scb_context scb_state;
 #endif
-	/* Set SLEEPDEEP bit of Cortex System Control Register */
-	LL_LPM_EnableDeepSleep();
-
-	while (LL_PWR_IsActiveFlag_ACTVOS() == 0) {
-	}
 
-	switch (substate_id) {
-	case 1: /* enter STOP0 mode */
-		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP0);
-		break;
-	case 2: /* enter STOP1 mode */
-		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP1);
-		break;
-	default:
-		LOG_DBG("Unsupported power state substate-id %u", substate_id);
-		break;
-	}
-}
+void stm32_power_init(void);
 
-#if defined(CONFIG_PM_S2RAM)
-static int suspend_to_ram(void)
+static int enter_low_power_mode(void)
 {
+	uint32_t basepri;
+
 	LL_LPM_EnableDeepSleep();
 
 	while (LL_PWR_IsActiveFlag_ACTVOS() == 0) {
 	}
 
-	/* Select mode entry : WFE or WFI and enter the CPU selected mode */
-	k_cpu_idle();
+	/*
+	 * Prevent spurious entry in low-power state if any
+	 * interrupt is currently pending. The WFI instruction
+	 * will not enter low-power state if an interrupt with
+	 * higher priority than BASEPRI is pending, regardless
+	 * of the value of PRIMASK.
+	 *
+	 * When this function executes, Zephyr's arch_irq_lock() has
+	 * configured BASEPRI to mask interrupts. Set PRIMASK to keep
+	 * interrupts masked then lower BASEPRI to 0, after saving its
+	 * current value, so that any pending interrupt is seen by WFI
+	 * and inhibits low-power state entry.
+	 */
+	__disable_irq();
+	basepri = __get_BASEPRI();
+	__set_BASEPRI(0);
+	__ISB();
+	__DSB();
+	__WFI();
 
-	return 0;
+	/*
+	 * If entry in STOP mode was attempted, execution
+	 * resumes here regardless of success or failure.
+	 * If entry in STANDBY mode was attempted, execution
+	 * will reach this point only if entry in low-power
+	 * state did not occur. Restore BASEPRI and clear
+	 * PRIMASK then return an error to indicate that
+	 * entry in STANDBY mode failed (return value of
+	 * this function is ignored for STOP mode entry).
+	 */
+	__set_BASEPRI(basepri);
+	__enable_irq();
+
```

Comment:
We also need to revert the `LL_LPM_EnableDeepSleep()`:
```suggestion
	__set_BASEPRI(basepri);
	__enable_irq();
	
	/* Disable SLEEPDEEP at Cortex-M level */
        LL_LPM_EnableSleep();
```